mscore/scoretreeview.cpp

```diff
@@ -0,0 +1,100 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#include "scoretreeview.h"
+
+namespace Ms {
+
+//---------------------------------------------------------
+//   selectElementAndChildren
+//---------------------------------------------------------
+
+static void selectElementAndChildren(Element* el)
+{
+    el->score()->selection().add(el);
+    for (ScoreElement* child : (*el)) {
+        if (child) {
+            selectElementAndChildren(toElement(child));
+        }
+    }
+}
+
+//---------------------------------------------------------
+//   ScoreTreeView
+//---------------------------------------------------------
+
+ScoreTreeView::ScoreTreeView(QWidget* parent)
+    : QTreeView(parent)
+{
+}
+
+//---------------------------------------------------------
+//   ~ScoreTreeView
+//---------------------------------------------------------
+
+ScoreTreeView::~ScoreTreeView()
+{
+}
+
+//---------------------------------------------------------
+//   setModel
+//---------------------------------------------------------
+
+void ScoreTreeView::setModel(QAbstractItemModel* model)
+{
+    Q_ASSERT(model);  // ensure there is a model
+    Q_ASSERT(dynamic_cast<ScoreItemModel*>(model));  // ensure model is correct type
+    QTreeView::setModel(model);
+}
+
+//---------------------------------------------------------
+//   model
+//---------------------------------------------------------
+
+ScoreItemModel* ScoreTreeView::scoreItemModel() const
+{
+    return static_cast<ScoreItemModel*>(QTreeView::model());
+}
+
+//---------------------------------------------------------
+//   currentChanged
+//---------------------------------------------------------
+
+void ScoreTreeView::currentChanged(const QModelIndex& current, const QModelIndex& previous)
+{
+    selectElementAtIndex(current);
+    QTreeView::currentChanged(current, previous);
+}
+
+//---------------------------------------------------------
+//   selectElementAtIndex
+//---------------------------------------------------------
+
+void ScoreTreeView::selectElementAtIndex(const QModelIndex& index)
+{
+    ScoreElement* el = static_cast<ScoreElement*>(index.internalPointer());
+
+    if (!el || !el->isElement()) {
+        return;
+    }
+
+    Score* cs = scoreItemModel()->score();
+    cs->selection().deselectAll();
+    cs->selection().setState(SelState::LIST);
+
+    selectElementAndChildren(toElement(el));
```

Comment:
We need a comment here to say [why it is necessary](https://github.com/musescore/MuseScore/pull/6174#discussion_r442144670) to select children of the element rather than just the element itself.

Another way around the leaf node problem would be to do a range selection.

```c++
if (el->treeChildCount() == 0) {
    // select el
} else {
    // select range that includes el
}
```

But let's not do that now. It may not be totally reliable anyway.